### PR TITLE
Fix rare corner cases in timer expiration

### PR DESCRIPTION
**Pull request by Keno**
See commit message for detailed description. Also fix a formatting issue in the log I noticed while debugging this.

### CODE REVIEW

**Comment by rocallahan** [comment]
Did you look at the failing `timer` test here?

**Comment by Keno** [comment]
I suspect it's a real timeout. This makes the `timer` test take quite a bit long and it would be even worse if context switches are slow (as they would be in a VM such as the buildbot). Not sure what to do. I already cut down the number of iterations in the test to a level where the original bug only reproduced every other run or so. If we cut down the iterations further, we probably risk the test not triggering what it's intended to test.

**Comment by rocallahan** [comment]
How about having the test check its own runtime and stop iterating if it runs for more than a minute? Then on a fast machine it should run to completion and on a slow machine it will do *some* testing and then exit gracefully if things are too slow.

**Comment by Keno** [comment]
Ok, now with self-monitoring execution time. We'll see what CI has to say.

**Comment by Keno** [comment]
Hmm, failed again. May actually have to log at the log. @khuey, emailing you an ssh key.

**Comment by Keno** [comment]
Not sure what's going on with the test failure. The time limiting seems to work fine outside rr on the buildbot (I tried setting it to 1sec), and it works fine under rr on my usual development machines.

**Comment by Keno** [comment]
Ah, got it. The context switch is so slow, that we never make it around the loop and the next timer expires before we get to check the flag. I'll have to have the overflow timer cancel the interval timer.

**Comment by Keno** [comment]
Ok, let's try that on the buildbot.

**Comment by Keno** [comment]
@rocallahan This should be ready for review as well, while we're at it.

File: src/test/timer.c [hunk 1]
```diff
@@ -4,61 +4,126 @@
...
 
-  for (counter = 0; counter >= 0 && !caught_sig; counter++) {
-    if (counter % 100000 == 0) {
-      write(STDOUT_FILENO, ".", 1);
+    /* This tries to reigger the following condition:
+     * - Timer expiration in user space
+     * - Right after a syscall return (same ip/ticks)
+     * The bug that this is a test for did not reproduce if the timer expired
+     * in kernel space or if there were any intervening ticks. In testing this
+     * took a few thousand iterrations to reproduce, so 5000 may not be
```
Comment (rocallahan) [inline src/test/timer.c]: "iterations"

File: src/test/timer.c [hunk 2]
```diff
@@ -4,61 +4,126 @@
...
+    sevp.sigev_notify = SIGEV_SIGNAL;
+    sevp.sigev_signo = SIGUSR1;
+    timer_create(CLOCK_MONOTONIC, &sevp, timeout_id);
+    caught_limit_sig = 0;
+    test_assert(0 == timer_settime(*timeout_id, 0, &timeout, NULL));
 
-  for (counter = 0; counter >= 0 && !caught_sig; counter++) {
-    if (counter % 100000 == 0) {
-      write(STDOUT_FILENO, ".", 1);
+    /* This tries to reigger the following condition:
```
Comment (rocallahan) [inline src/test/timer.c]: "trigger"

**Comment by rocallahan** [comment]
Only those trivial typos, the rest looks fine.

**Comment by Keno** [comment]
Thanks, fixed.

**Comment by Keno** [comment]
Bump

**Comment by rocallahan** [comment]
Sorry!

**Pull request merged**